File: TimeKeeperAndroid/src/com/tjoris/timekeeper/PlaylistEditActivity.java
package com.tjoris.timekeeper;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Dialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AbsListView;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemClickListener;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.SimpleAdapter;

import com.tjoris.timekeeper.data.IPlaylistStore;
import com.tjoris.timekeeper.data.Playlist;
import com.tjoris.timekeeper.data.PlaylistStoreFactory;
import com.tjoris.timekeeper.data.Song;

public class PlaylistEditActivity extends Activity
{
	private static final String kKEY_NAME = "name";
	private static final String kKEY_TEMPO = "tempo";

	private final List<Map<String, String>> fData;
	private InputDialog fAddSongDialog;
	private InputDialog fRenamePlaylistDialog;
	private InputDialog fCopyDialog;
	private InputDialog fEditSongDialog;
	private final ConfirmationDialog fDeleteDialog;
	private final IPlaylistStore fStore;
	private Playlist fPlaylist = null;
	private ActionMode fActionMode = null;
	private int fPosition;

	public PlaylistEditActivity()
	{
		fData = new ArrayList<Map<String, String>>();
		fStore = PlaylistStoreFactory.createStore(this);

		fDeleteDialog = new ConfirmationDialog()
		{
			@Override
			public void confirm()
			{
				deleteSelectedItems();
			}
		};
		fDeleteDialog.setOptions(R.string.playlistedit_delete_message, R.string.playlistedit_delete_yes, R.string.playlistedit_delete_no);
	}

	@SuppressLint("ClickableViewAccessibility")
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		fAddSongDialog = new InputDialog(getLayoutInflater(),
				R.string.playlistedit_action_addsong,
				R.drawable.ic_action_new,
				R.layout.playlistedit_song,
				R.string.playlistedit_addsong_add,
				R.string.playlistedit_addsong_cancel
		)
		{
			@Override
			public void dialogConfirmed(final Dialog dialog)
			{
				final CharSequence name = ((EditText)dialog.findViewById(R.id.playlistedit_song_name)).getText();
				final CharSequence tempoS = ((EditText)dialog.findViewById(R.id.playlistedit_song_tempo)).getText();
				final int tempo = parseTempo(tempoS);
				final Song song = new Song(name.toString(), tempo);
				fPlaylist.addSong(fStore, song);
				reloadSongs();
			}
		};
		fRenamePlaylistDialog = new InputDialog(getLayoutInflater(),
				R.string.playlistedit_action_renameplaylist,
				R.drawable.ic_action_edit,
				R.layout.playlistedit_renameplaylist,
				R.string.playlistedit_renameplaylist_save,
				R.string.playlistedit_renameplaylist_cancel
		)
		{
			@Override
			public void viewCreated(final View view)
			{
				((EditText)view.findViewById(R.id.playlistedit_renameplaylist_name)).setText(fPlaylist != null ? fPlaylist.getName() : "");
			}
			
			@Override
			public void dialogConfirmed(final Dialog dialog)
			{
				final CharSequence name = ((EditText)dialog.findViewById(R.id.playlistedit_renameplaylist_name)).getText();
				fPlaylist.setName(fStore, name.toString());
				reloadName();
			}
		};
		fCopyDialog = new InputDialog(getLayoutInflater(),
				R.string.playlistedit_action_copy,
				R.drawable.ic_action_copy,
				R.layout.playlistedit_copy,
				R.string.playlistedit_copy_copy,
				R.string.playlistedit_copy_cancel
		)
		{
			@Override
			public void viewCreated(final View view)
			{
				((EditText)view.findViewById(R.id.playlistedit_copy_name)).setText(fPlaylist != null ? fPlaylist.getName() : "");
			}
			
			@Override
			public void dialogConfirmed(final Dialog dialog)
			{
				final CharSequence name = ((EditText)dialog.findViewById(R.id.playlistedit_copy_name)).getText();
				final Playlist playlist = new Playlist(fPlaylist, name.toString(), fStore.getNextPlaylistWeight());
				fStore.addPlaylist(playlist);
				
				final Intent intent = new Intent(PlaylistEditActivity.this, PlaylistActivity.class);
				intent.putExtra("playlist", playlist.getID());
				startActivity(intent);
			}
		};
		fEditSongDialog = new InputDialog(getLayoutInflater(),
				R.string.playlistedit_action_editsong,
				R.drawable.ic_action_edit,
				R.layout.playlistedit_song,
				R.string.playlistedit_editsong_save,
				R.string.playlistedit_editsong_cancel
		)
		{
			@Override
			public void viewCreated(final View view)
			{
				final Song song = fPlaylist.getSongs().get(fPosition);
				((EditText)view.findViewById(R.id.playlistedit_song_name)).setText(song.getName());
				((EditText)view.findViewById(R.id.playlistedit_song_tempo)).setText(Integer.toString(song.getTempo()));
			}
			
			@Override
			public void dialogConfirmed(final Dialog dialog)
			{
				final Song song = fPlaylist.getSongs().get(fPosition);
				final CharSequence name = ((EditText)dialog.findViewById(R.id.playlistedit_song_name)).getText();
				final CharSequence tempoS = ((EditText)dialog.findViewById(R.id.playlistedit_song_tempo)).getText();
				final int tempo = parseTempo(tempoS);
				song.update(fStore, fPlaylist, name.toString(), tempo);
				reloadSongs();
			}
		};

		setContentView(R.layout.playlistedit);
		getActionBar().setDisplayHomeAsUpEnabled(true);

		final ListView playlistView = getPlaylist();
		playlistView.setAdapter(new SimpleAdapter(this, fData, R.layout.playlist_entry, new String[] { kKEY_NAME, kKEY_TEMPO }, new int[] { R.id.playlist_entry_name, R.id.playlist_entry_tempo }));
		playlistView.setOnItemClickListener(new OnItemClickListener()
		{
			@Override
			public void onItemClick(final AdapterView<?> parent, final View view, final int position, final long id)
			{
				if (fPlaylist != null && position >= 0 && position < fPlaylist.getSongs().size())
				{
					fPosition = position;
					fEditSongDialog.show(getFragmentManager(), "edit_song");
				}
			}
		});
		playlistView.setMultiChoiceModeListener(new AbsListView.MultiChoiceModeListener()
		{
			@Override
			public boolean onPrepareActionMode(final ActionMode mode, final Menu menu)
			{
				return false;
			}

			@Override
			public void onDestroyActionMode(final ActionMode mode)
			{
				fActionMode = null;
			}

			@Override
			public boolean onCreateActionMode(final ActionMode mode, final Menu menu)
			{
				fActionMode = mode;
				mode.getMenuInflater().inflate(R.menu.playlistedit_contextactions, menu);
				return true;
			}

			@Override
			public boolean onActionItemClicked(final ActionMode mode, final MenuItem item)
			{
				switch (item.getItemId())
				{
				case R.id.playlistedit_action_up:
				{
					move(true);
					return true;
				}
				case R.id.playlistedit_action_down:
				{
					move(false);
					return true;
				}
				case R.id.playlistedit_action_delete:
				{
					fDeleteDialog.show(getFragmentManager(), "delete_songs");
					return true;
				}
				default:
				{
					return false;
				}
				}
			}

			@Override
			public void onItemCheckedStateChanged(final ActionMode mode, final int position, final long id, final boolean checked)
			{
			}
		});
		
		loadIntent();
	}
	
	private void loadIntent()
	{
		final Bundle bundle = getIntent().getExtras();
		if (bundle != null)
		{
			final long playlist = bundle.getLong("playlist");
			if (fPlaylist == null || fPlaylist.getID() != playlist)
			{
				fPlaylist = fStore.readPlaylist(playlist);
				loadPlaylist();
			}
		}
		else
		{
			fPlaylist = null;
			loadPlaylist();
		}
	}

	@Override
	protected void onDestroy()
	{
		fStore.close();
		super.onDestroy();
	}

	@Override
	public boolean onCreateOptionsMenu(final Menu menu)
	{
		getMenuInflater().inflate(R.menu.playlistedit_actions, menu);
		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(final MenuItem item)
	{
		// Handle presses on the action bar items
		switch (item.getItemId())
		{
		case R.id.playlistedit_action_addsong:
		{
			fAddSongDialog.show(getFragmentManager(), "addsong");
			return true;
		}
		case R.id.playlistedit_action_renameplaylist:
		{
			fRenamePlaylistDialog.show(getFragmentManager(), "renameplaylist");
			return true;
		}
		case R.id.playlistedit_action_copy:
		{
			fCopyDialog.show(getFragmentManager(), "copyplaylist");
			return true;
		}
		default:
		{
			return super.onOptionsItemSelected(item);
		}
		}
	}

	private void loadPlaylist()
	{
		reloadName();
		reloadSongs();
	}
	
	private void reloadName()
	{
		if (fPlaylist != null)
		{
			getActionBar().setTitle(fPlaylist.getName());
		}
		else
		{
			getActionBar().setTitle("<no playlist>");
		}
	}
	
	private void reloadSongs()
	{
		fData.clear();
		if (fPlaylist != null)
		{
			for (final Song song : fPlaylist.getSongs())
			{
				final Map<String, String> map = new HashMap<String, String>();
				map.put(kKEY_NAME, song.getName());
				map.put(kKEY_TEMPO, Integer.toString(song.getTempo()));
				fData.add(map);
			}
		}
		((SimpleAdapter)getPlaylist().getAdapter()).notifyDataSetChanged();
	}

	private void deleteSelectedItems()
	{
		final ListView playlist = getPlaylist();
		final SparseBooleanArray selection = playlist.getCheckedItemPositions();
		for (int i = fData.size() - 1; i >= 0; --i)
		{
			if (selection.get(i))
			{
				fPlaylist.removeSong(fStore, i);
			}
		}
		if (fActionMode != null)
		{
			fActionMode.finish();
		}
		reloadSongs();
	}

	private void move(final boolean up)
	{
		final ListView playlist = getPlaylist();
		final SparseBooleanArray selection = playlist.getCheckedItemPositions();
		boolean ignore = false;
		for (int i = up ? 0 : fData.size() - 1; up ? i < fData.size() : i >= 0; i += up ? 1 : -1)
		{
			if (selection.get(i))
			{
				if (up ? i <= 0 : i >= fData.size() - 1)
				{
					// leave the first selection group alone
					ignore = true;
				}
				else if (!ignore)
				{
					final int otherIndex = up ? i - 1 : i + 1;
					if (!selection.get(otherIndex))
					{
						playlist.setItemChecked(i, false);
						playlist.setItemChecked(otherIndex, true);
					}
					fPlaylist.move(fStore, i, up);
				}
			}
			else
			{
				ignore = false;
			}
		}
		reloadSongs();
	}
	
	private ListView getPlaylist()
	{
		return (ListView) findViewById(R.id.playlistedit);
	}
	
	private static int parseTempo(final CharSequence tempo)
	{
		try
		{
			return Math.min(300, Math.max(30, Integer.parseInt(tempo.toString())));
		}
		catch (final NumberFormatException e)
		{
			Log.e("TimeKeeper", "Invalid tempo: " + e.getMessage(), e);
		}
		return 120;
	}
}

File: TimeKeeperAndroid/app/src/main/java/be/t_ars/timekeeper/PlaylistEditActivity.kt
package be.t_ars.timekeeper

import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.ActionMode
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AbsListView
import android.widget.AdapterView
import android.widget.EditText
import android.widget.SimpleAdapter
import be.t_ars.timekeeper.components.PlaylistList
import be.t_ars.timekeeper.components.SongList
import be.t_ars.timekeeper.data.*
import be.t_ars.timekeeper.databinding.PlaylisteditBinding
import kotlin.math.max
import kotlin.math.min


private const val kKEY_NAME = "name"
private const val kKEY_TEMPO = "tempo"

private fun parseTempo(tempo: CharSequence): Int {
    if (tempo.isBlank()) {
        return ClickDescription.DEFAULT_TEMPO
    }
    try {
        return min(300, max(30, Integer.parseInt(tempo.toString())))
    } catch (e: NumberFormatException) {
        Log.e("TimeKeeper", "Invalid tempo: " + e.message, e)
    }
    return ClickDescription.DEFAULT_TEMPO
}

private enum class EditMode {
    NORMAL, SEND_TOP, SEND_BOTTOM
}

class PlaylistEditActivity : AbstractActivity() {
    private lateinit var fBinding: PlaylisteditBinding
    private val fData: MutableList<Map<String, String>> = ArrayList()
    private val fAddSongDialog: InputDialog = InputDialog()
    private val fRenamePlaylistDialog: InputDialog = InputDialog()
    private val fCopyDialog: InputDialog = InputDialog()
    private val fDeleteDialog: ConfirmationDialog = ConfirmationDialog()
    private val fDeleteAndAboveDialog: ConfirmationDialog = ConfirmationDialog()
    private val fInsertfromPlaylistList = PlaylistList()
    private val fInsertfromSongList = SongList()

    private lateinit var fStore: PlaylistStore
    private var fPlaylist: Playlist? = null
    private var fActionMode: ActionMode? = null
    private var fPosition: Int = 0
    private var fNewPlaylistId: Long? = null
    private var fEditMode = EditMode.NORMAL
    private val fInsertFromPlaylists: MutableList<PlaylistHeader> = ArrayList()
    private val fInsertFromSongs: MutableList<Song> = ArrayList()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        fStore = PlaylistStore(this)

        fAddSongDialog.setOptions(
            null,
            { view ->
                val name = view.findViewById<EditText>(R.id.playlistedit_song_name).text
                val tempoS = view.findViewById<EditText>(R.id.playlistedit_song_tempo).text
                val tempo = parseTempo(tempoS)
                val song =
                    Song(
                        name.toString(),
                        ClickDescription(
                            tempo,
                            EClickType.DEFAULT,
                            ClickDescription.DEFAULT_DIVISION_COUNT,
                            ClickDescription.DEFAULT_BEAT_COUNT,
                            ClickDescription.DEFAULT_COUNT_OFF
                        )
                    )
                fPlaylist?.addSong(fStore, song)
                reloadSongs()
            },
            layoutInflater,
            R.string.playlistedit_action_addsong,
            R.drawable.ic_plus_dark,
            R.layout.playlistedit_song,
            R.string.add,
            R.string.cancel
        )
        fRenamePlaylistDialog.setOptions(
            { view ->
                view.findViewById<EditText>(R.id.playlistedit_renameplaylist_name)
                    .setText(fPlaylist?.name ?: "")
            },
            { view ->
                val name = view.findViewById<EditText>(R.id.playlistedit_renameplaylist_name).text
                fPlaylist?.let {
                    it.name = name.toString()
                    fStore.storePlaylistHeader(it)
                    reloadName()
                }
            },
            layoutInflater,
            R.string.playlistedit_action_renameplaylist,
            R.drawable.ic_pencil,
            R.layout.playlistedit_renameplaylist,
            R.string.playlistedit_renameplaylist_save,
            R.string.playlistedit_renameplaylist_cancel
        )
        fCopyDialog.setOptions(
            { view ->
                view.findViewById<EditText>(R.id.playlistedit_copy_name).setText(
                    fPlaylist?.name
                        ?: ""
                )
            },
            { view -> copyPlaylist(view.findViewById<EditText>(R.id.playlistedit_copy_name).text.toString()) },
            layoutInflater,
            R.string.playlistedit_action_copy,
            R.drawable.ic_content_copy,
            R.layout.playlistedit_copy,
            R.string.playlistedit_copy_copy,
            R.string.playlistedit_copy_cancel
        )
        fDeleteDialog.setOptions(
            this::deleteSelectedItems,
            R.string.playlistedit_delete_message,
            R.string.playlistedit_delete_yes,
            R.string.playlistedit_delete_no
        )
        fDeleteAndAboveDialog.setOptions(
            this::deleteSelectedItemsAndAbove,
            R.string.playlistedit_delete_and_above_message,
            R.string.playlistedit_delete_yes,
            R.string.playlistedit_delete_no
        )

        fBinding = PlaylisteditBinding.inflate(layoutInflater)
        setContentView(fBinding.root)
        setSupportActionBar(fBinding.toolbar)

        fBinding.fabAddSong.setOnClickListener {
            fAddSongDialog.show(supportFragmentManager, "addsong")
        }

        fBinding.playlistedit.adapter = SimpleAdapter(
            this,
            fData,
            R.layout.playlist_entry,
            arrayOf(kKEY_NAME, kKEY_TEMPO),
            intArrayOf(R.id.playlist_entry_name, R.id.playlist_entry_tempo)
        )
        fBinding.playlistedit.onItemClickListener =
            AdapterView.OnItemClickListener { _, _, position, _ ->
                if (fPlaylist != null && position >= 0 && position < (fPlaylist?.songs?.size
                        ?: 0)
                ) {
                    when (fEditMode) {
                        EditMode.NORMAL -> {
                            fPosition = position
                            editEntry()
                        }

                        EditMode.SEND_TOP -> {
                            fPlaylist?.sendToTop(fStore, position)
                            reloadSongs()
                        }

                        EditMode.SEND_BOTTOM -> {
                            fPlaylist?.sendToBottom(fStore, position)
                            reloadSongs()
                        }
                    }
                }
            }

        fBinding.playlistedit.setMultiChoiceModeListener(object :
            AbsListView.MultiChoiceModeListener {
            override fun onPrepareActionMode(mode: ActionMode, menu: Menu): Boolean {
                return false
            }

            override fun onDestroyActionMode(mode: ActionMode) {
                fActionMode = null
            }

            override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
                fActionMode = mode
                mode.menuInflater.inflate(R.menu.playlistedit_contextactions, menu)
                return true
            }

            override fun onActionItemClicked(mode: ActionMode, item: MenuItem): Boolean {
                when (item.itemId) {
                    R.id.playlistedit_action_up -> {
                        move(true)
                    }

                    R.id.playlistedit_action_down -> {
                        move(false)
                    }

                    R.id.playlistedit_action_delete -> {
                        fDeleteDialog.show(supportFragmentManager, "delete_songs")
                    }

                    R.id.playlistedit_action_delete_and_above -> {
                        fDeleteAndAboveDialog.show(supportFragmentManager, "delete_songs_and_above")
                    }

                    else -> {
                        return false
                    }
                }
                return true
            }

            override fun onItemCheckedStateChanged(
                mode: ActionMode,
                position: Int,
                id: Long,
                checked: Boolean
            ) {
            }
        })

        fBinding.playlisteditSelectfromClose.setOnClickListener {
            fBinding.playlisteditSelectfromLayout.visibility = View.INVISIBLE
        }
    }

    private fun copyPlaylist(name: String) {
        fPlaylist?.let { p ->
            fNewPlaylistId?.let { id ->
                val newPlaylist = Playlist(p, id, name, fStore.nextPlaylistWeight)
                fStore.addPlaylist(newPlaylist)
                fStore.setCurrentPlaylistID(newPlaylist.id)

                PlaylistActivity.startActivity(this)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        requestedOrientation = getSettingScreenOrientation(this)
        loadIntent()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            kREQUEST_TEMPO -> {
                if (resultCode != RESULT_OK) {
                    return
                }
                data?.let { d ->
                    fPlaylist?.let { playlist ->
                        if (fPosition in 0 until playlist.songs.size) {
                            val newName = d.getStringExtra(TapSongActivity.kINTENT_DATA_NAME)
                            val newClick =
                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
                                    d.getSerializableExtra(
                                        TapSongActivity.kINTENT_DATA_CLICK,
                                        ClickDescription::class.java
                                    )
                                else
                                    d.getSerializableExtra(TapSongActivity.kINTENT_DATA_CLICK) as ClickDescription
                            val newScoreLink =
                                d.getStringExtra(TapSongActivity.kINTENT_DATA_SCORE_LINK)
                                    ?.let { it.ifBlank { null } }

                            val song = playlist.songs[fPosition]
                            val replaceName = newName != null && newName != song.name
                            val replaceClick = newClick != song.click
                            val replaceScoreLink = newScoreLink != song.scoreLink
                            if (replaceName || replaceClick || replaceScoreLink) {
                                if (replaceName && newName != null)
                                    song.name = newName
                                if (replaceClick && newClick != null)
                                    song.click = newClick
                                if (replaceScoreLink)
                                    song.scoreLink = newScoreLink
                                fStore.savePlaylist(playlist)
                                reloadSongs()
                            }
                        }
                    }
                }
            }

            kREQUEST_DOCUMENT_CODE -> fStore.acceptRequestedDocument(data) { id ->
                fNewPlaylistId = id
                val ft = supportFragmentManager.beginTransaction()
                ft.add(fCopyDialog, null)
                ft.commitAllowingStateLoss()
            }

            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.playlistedit_actions, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle presses on the action bar items
        when (item.itemId) {
            R.id.playlistedit_action_renameplaylist -> {
                fRenamePlaylistDialog.show(supportFragmentManager, "renameplaylist")
            }

            R.id.playlistedit_action_copy -> {
                startActivityForResult(fStore.createDocumentRequest(), kREQUEST_DOCUMENT_CODE)
            }

            R.id.playlistedit_action_insertfrom -> {
                startInsertFrom()
            }

            R.id.playlistedit_action_normalmode -> {
                fEditMode = EditMode.NORMAL
                reloadSongs()
            }

            R.id.playlistedit_action_sendtop -> {
                fEditMode = EditMode.SEND_TOP
                reloadSongs()
            }

            R.id.playlistedit_action_sendbottom -> {
                fEditMode = EditMode.SEND_BOTTOM
                reloadSongs()
            }

            else -> {
                return super.onOptionsItemSelected(item)
            }
        }
        return true
    }

    private fun loadIntent() {
        val bundle = intent.extras
        if (bundle != null) {
            val playlist = bundle.getLong(kINTENT_DATA_PLAYLIST_ID)
            val p = fPlaylist
            if (p == null || p.id != playlist) {
                fPlaylist = fStore.readPlaylist(playlist)
                loadPlaylist()
            }
        } else {
            fPlaylist = null
            loadPlaylist()
        }
    }

    private fun loadPlaylist() {
        fEditMode = EditMode.NORMAL
        reloadName()
        reloadSongs()
    }

    private fun editEntry() {
        fPlaylist?.let { playlist ->
            val song = playlist.songs[fPosition]
            TapSongActivity.startActivityForResult(
                this,
                song.click,
                song.name,
                song.scoreLink,
                kREQUEST_TEMPO
            )
        }
    }

    private fun reloadName() {
        supportActionBar?.title = fPlaylist?.name ?: "<no playlist>"
    }

    private fun reloadSongs() {
        fData.clear()
        fPlaylist?.let { playlist ->
            fData.addAll(
                playlist.songs.map { song ->
                    var name = song.displayName()
                    when (fEditMode) {
                        EditMode.SEND_TOP -> name = "\u2912 $name"
                        EditMode.SEND_BOTTOM -> name = "\u2913 $name"
                        else -> {}
                    }
                    val tempo = "${song.click.bpm}"
                    mapOf(kKEY_NAME to name, kKEY_TEMPO to tempo)
                }
            )
        }
        (fBinding.playlistedit.adapter as SimpleAdapter).notifyDataSetChanged()
    }

    private fun deleteSelectedItems() {
        val selection = fBinding.playlistedit.checkedItemPositions
        fData.indices.reversed().forEach { i ->
            if (selection.get(i)) {
                fPlaylist?.removeSong(fStore, i)
            }
        }
        fActionMode?.finish()
        reloadSongs()
    }

    private fun deleteSelectedItemsAndAbove() {
        val selection = fBinding.playlistedit.checkedItemPositions
        val highestSelection = fData.indices.reversed().firstOrNull(selection::get)
        if (highestSelection != null) {
            fPlaylist?.removeSongAndAbove(fStore, highestSelection)
            fActionMode?.finish()
            reloadSongs()
        }
    }

    private fun move(up: Boolean) {
        val selection = fBinding.playlistedit.checkedItemPositions
        var ignore = false
        var i = if (up) 0 else fData.size - 1
        while (if (up) i < fData.size else i >= 0) {
            if (selection.get(i)) {
                if (if (up) i <= 0 else i >= fData.size - 1) {
                    // leave the first selection group alone
                    ignore = true
                } else if (!ignore) {
                    val otherIndex = if (up) i - 1 else i + 1
                    if (!selection.get(otherIndex)) {
                        fBinding.playlistedit.setItemChecked(i, false)
                        fBinding.playlistedit.setItemChecked(otherIndex, true)
                    }
                    fPlaylist?.move(fStore, i, up)
                }
            } else {
                ignore = false
            }
            i += if (up) 1 else -1
        }
        reloadSongs()
    }

    private fun startInsertFrom() {
        fInsertfromPlaylistList.bindToView(this, fBinding.playlisteditSelectfromList)

        fInsertFromPlaylists.clear()
        fInsertFromPlaylists.addAll(fStore.readAllPlaylists())
        fInsertfromPlaylistList.setPlaylists(fInsertFromPlaylists)

        fBinding.playlisteditSelectfromLayout.visibility = View.VISIBLE

        fBinding.playlisteditSelectfromList.onItemClickListener =
            AdapterView.OnItemClickListener { _, _, position, _ ->
                loadInsertFromPlaylist(position)
            }
    }

    private fun loadInsertFromPlaylist(selection: Int) {
        val playlist = fStore.readPlaylist(fInsertFromPlaylists[selection].id)
        if (playlist != null) {
            fInsertfromSongList.bindToView(this, fBinding.playlisteditSelectfromList)

            fInsertFromSongs.clear()
            fInsertFromSongs.addAll(playlist.songs)
            fInsertfromSongList.setSongs(fInsertFromSongs)

            fBinding.playlisteditSelectfromList.onItemClickListener =
                AdapterView.OnItemClickListener { _, _, position, _ ->
                    fBinding.playlisteditSelectfromList.setItemChecked(position, true)

                    val song = fInsertFromSongs[position]
                    fPlaylist?.addSong(fStore, Song(song))
                    reloadSongs()

                    fPlaylist?.songs?.size?.also {
                        val pos = it - 1
                        //fBinding.playlistedit.setItemChecked(pos, true)
                        fBinding.playlistedit.smoothScrollToPosition(pos)
                    }
                }
        }
    }

    companion object {
        private const val kINTENT_DATA_PLAYLIST_ID = "playlist-id"

        private const val kREQUEST_TEMPO = 1
        private const val kREQUEST_DOCUMENT_CODE = 2

        fun startActivity(context: Context, playlistID: Long) =
            Intent(context, PlaylistEditActivity::class.java)
                .also { intent ->
                    intent.putExtra(kINTENT_DATA_PLAYLIST_ID, playlistID)
                }
                .let(context::startActivity)
    }
}
